### PR TITLE
Implement OAuth over S2S via CopySourceAuthorization

### PR DESCRIPTION
**Pull request by adreed-msft**
This PR is dependent upon https://github.com/Azure/azure-storage-blob-go/pull/317.

Makes use of the new CopySourceAuthorization header available on (most) S2S operations. Additionally, this PR splits up the STE's understanding of auth types into primary and S2S source auth types. Whenever the OAuth token may be needed on either the source or the destination, it is sent into the STE.

### CODE REVIEW

File: cmd/copyEnumeratorInit.go [hunk 1]
```diff
@@ -42,11 +42,29 @@ func (cca *CookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrde
...
+	} else if cca.FromTo.IsS2S() &&
+		((srcCredInfo.CredentialType == common.ECredentialType.OAuthToken() && cca.FromTo.To() != common.ELocation.Blob()) || // Blob can forward OAuth tokens
 			(srcCredInfo.CredentialType == common.ECredentialType.Anonymous() && !isPublic && cca.Source.SAS == "")) {
-		// TODO: Generate a SAS token if it's blob -> *
-		return nil, errors.New("a SAS token (or S3 access key) is required as a part of the source in S2S transfers, unless the source is a public resource")
+		return nil, errors.New("a SAS token (or S3 access key) is required as a part of the source in S2S transfers, unless the source is a public resource, or the destination is blob storage")
+	}
+
+	if cca.Source.SAS != "" && cca.FromTo.IsS2S() && jobPartOrder.CredentialInfo.CredentialType == common.ECredentialType.OAuthToken() {
+		glcm.Info("SECURITY: If the source and destination accounts are in the same AAD tenant & the OAuth token has appropriate permissions on both, the source SAS token is not required and OAuth can be used round-trip.")
```
Comment (zezha-msft) [inline cmd/copyEnumeratorInit.go:52]: Clarify that it's the user/spn/msi that must have the appropriate permission, since that's what the user controls

Comment (zezha-msft) [reply cmd/copyEnumeratorInit.go:52]: SECURITY:  is kind of alarming

File: common/rpc-models.go [hunk 2]
```diff
@@ -148,6 +148,9 @@ type CopyJobPartOrderRequest struct {
 	S2SInvalidMetadataHandleOption InvalidMetadataHandleOption
 	S2SPreserveBlobTags            bool
 	CpkOptions                     CpkOptions
+
+	PrimaryCredentialType   CredentialType
```
Comment (zezha-msft) [inline common/rpc-models.go:152]: The naming is a bit confusing, how do these two credential types interact with each other?

File: go.mod [hunk 3]
```diff
@@ -3,8 +3,8 @@ module github.com/Azure/azure-storage-azcopy/v10
 require (
 	cloud.google.com/go/iam v0.3.0 // indirect
 	cloud.google.com/go/storage v1.21.0
-	github.com/Azure/azure-pipeline-go v0.2.3
-	github.com/Azure/azure-storage-blob-go v0.13.1-0.20210914164749-2d6cd3e07548
+	github.com/Azure/azure-pipeline-go v0.2.4-0.20220420205509-9c760f3e9499
+	github.com/Azure/azure-storage-blob-go v0.13.1-0.20220418220008-28ac0a48144e
```
Comment (zezha-msft) [inline go.mod:7]: update to released version

**Review by zezha-msft — COMMENTED** [review]
Could you please give a description of how these 3 interact with each other?

CredentialInfo:          order.CredentialInfo,
PrimaryCredentialType:   order.PrimaryCredentialType,
S2SSourceCredentialType: order.S2SSourceCredentialType,

File: e2etest/zt_client_provided_key_test.go [hunk 4]
```diff
@@ -36,116 +30,116 @@ import (
...
-	}, nil, testFiles{
-		defaultSize: "100K",
-		shouldTransfer: []interface{}{
-			f("file1", verifyOnlyProps),
-			folder("dir"),
-			f("dir/file2", verifyOnlyProps),
-		},
-	}, EAccountType.Standard(), EAccountType.Standard(), "")
-}
+// func TestClient_ProvidedScopeUpload(t *testing.T) {
```
Comment (zezha-msft) [inline e2etest/zt_client_provided_key_test.go:33]: uncomment

**Review by siminsavani-msft — APPROVED** [review]

**Review by zezha-msft — APPROVED** [review]

**Pull request merged**